### ISSUE TITLE
sage-notebook-0.11.4-r2.ebuild fails install phase in prefix [double prefix]

### ISSUE THREAD

**Issue by strogdon** [cschwan/sage-on-gentoo#345]
In prefix with the subject notebook I get
```
 installing sage-notebook-0.11.4-r2 into /local/gentoo-redlizard/var/tmp/portage/sci-mathematics/sage-notebook-0.11.4-r2/image/

 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/ double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr/lib double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr/lib/python2.7 double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr/lib/python2.7/site-packages double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data double prefix
 ^[[33;01m*^[[0m QA Notice: local/gentoo-redlizard///local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax double prefix
 ^[[31;01m*^[[0m ERROR: sci-mathematics/sage-notebook-0.11.4-r2::sage-on-gentoo failed:
 ^[[31;01m*^[[0m   Aborting due to QA concerns: double prefix files installed
 ^[[31;01m*^[[0m
 ^[[31;01m*^[[0m Call stack:
 ^[[31;01m*^[[0m   misc-functions.sh, line 1191:  Called install_qa_check
 ^[[31;01m*^[[0m   misc-functions.sh, line  187:  Called source 'install_symlink_html_docs'
 ^[[31;01m*^[[0m            05prefix, line  115:  Called install_qa_check_prefix
 ^[[31;01m*^[[0m            05prefix, line   19:  Called die
 ^[[31;01m*^[[0m The specific snippet of code:
 ^[[31;01m*^[[0m                die "Aborting due to QA concerns: double prefix files installed"
```
At issue seems to be the mathjax links
```
        # link in system mathjax
        dosym /usr/share/mathjax "$(python_get_sitedir)"/sagenb/data/mathjax
}

pkg_preinst() {
        # remove old mathjax folder if present
        [[ -d "${ROOT}$(python_get_sitedir)/sagenb/data/mathjax" ]] \
                && rm -rf "${ROOT}$(python_get_sitedir)/sagenb/data/mathjax"
}
```
The $(python_get_sitedir) variable doesn't have EPREFIX included. But even including it does not fix things.

**Comment by strogdon** [comment]
That's not what I meant to say. $(python_get_sitedir) does include EPREFIX and ``stripping`` EPREFIX doesn't fix things. Stripping does seem to place things in the correct locations under the /image folder but there is
```
grep: /local/gentoo-redlizard/var/tmp/portage/sci-mathematics/sage-notebook-0.11.4-r2/image/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax: Is a directory
ecompressdir: bzip2 -9 /usr/share/doc
```

**Comment by kiwifb** [comment]
Can you try replacing 
```
dosym /usr/share/mathjax "$(python_get_sitedir)"/sagenb/data/mathjax
```
with 
```
ln -s ${EPREFIX}/usr/share/mathjax "$(python_get_sitedir)"/sagenb/data/mathjax
```
I have had serious troubles using dosym will working on issue #344.

**Comment by strogdon** [comment]
That doesn't work. Still double prefix. This seems to work
```
        SITEDIR=$(python_get_sitedir)
        # link in system mathjax
        dosym /usr/share/mathjax "${SITEDIR#${EPREFIX}}"/sagenb/data/mathjax
```
but there is the
```
grep: /local/gentoo-redlizard/var/tmp/portage/sci-mathematics/sage-notebook-0.11.4-r2/image/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax: Is a directory
```
and I don't know from where it is coming. I don't see it on native Gentoo.

**Comment by strogdon** [comment]
I missed the hard creation of the symlink in your last post. Let me try that.

**Comment by strogdon** [comment]
OK, with
```
ln -s ${EPREFIX}/usr/share/mathjax "$(python_get_sitedir)"/sagenb/data/mathjax
```
there is
```
 * python2_7: running distutils-r1_run_phase distutils-r1_python_install_all
ln: failed to create symbolic link ‘/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax/mathjax’: File exists
>>> Completed installing sage-notebook-0.11.4-r2 into /local/gentoo-redlizard/var/tmp/portage/sci-mathematics/sage-notebook-0.11.4-r2/image/
```
So why the extra ``mathjax``? Is it because of the already-installed notebook?

**Comment by kiwifb** [comment]
I would like you to unmerge sage-notebook and then comb ‘/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb` to see if there are still any files there that shouldn't be. It is possible you have file not belonging to any ebuild that were created by problem in a previous run.

**Comment by strogdon** [comment]
Yes there was the symlink
```
/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax/mathjax -> /local/gentoo-redlizard/usr/share/mathjax
```
So after removing the symlink and with sage-notebook-0.11.4 installed when I 
```
ebuild sage-notebook-0.11.4-r2.ebuild clean install
```
the
```
ln -s ${EPREFIX}/usr/share/mathjax "$(python_get_sitedir)"/sagenb/data/mathjax
```
recreates the undesirable symlink. Without any notebook installed there is
```
 * python2_7: running distutils-r1_run_phase distutils-r1_python_install_all
ln: failed to create symbolic link ‘/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax’: No such file or directory
```
So the ebuild is trying to create the symlink in the ``live`` system which I don't think is good?

**Comment by kiwifb** [comment]
Indeed it isn't. So let's change things a bit. We link before `distutils-r1_src_install` and with
`ln -s ${EPREFIX}/usr/share/mathjax sagenb/data/mathjax`

**Comment by kiwifb** [comment]
Actually I am not sure why I didn't follow the model from the `ipython` ebuild exactly:
```
	distutils-r1_src_install
	ln -snf "${EPREFIX}"/usr/share/mathjax \
		"${D}$(python_get_sitedir)"/sagenb/data/mathjax || die
```
Surely, if it works for ipython it should work for sage-notebook.

**Comment by strogdon** [comment]
This seems to work however there is still
```
grep: /local/gentoo-redlizard/var/tmp/portage/sci-mathematics/sage-notebook-0.11.4-r2/image/local/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax: Is a directory
```
which may just be a prefix thing? It doesn't prevent the install. There must be some reason for doing the above although
```
        SITEDIR=$(python_get_sitedir)
        # link in system mathjax
        dosym /usr/share/mathjax "${SITEDIR#${EPREFIX}}"/sagenb/data/mathjax
```
seems to give the same result. The ipython result seems cleaner.

**Comment by kiwifb** [comment]
I'll commit the `ipython` way after my lunch. I have to fix all those things in 6.7.beta, I just hope beta1 doesn't break the stuff I already done for beta0 in baselayout. I cannot complain I was expecting most of these changes since I reviewed them.

**Comment by strogdon** [comment]
I don't believe the following is prefix specific, but on upgrading from an older notebook to `sage-notebook-0.11.4-r2` I see
```
>>> Installing (5 of 5) sci-mathematics/sage-notebook-0.11.4-r2::sage-on-gentoo
 * 
 * Installation of a symlink is blocked by a directory:
 *   '/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax'
 * This symlink will be merged with a different name:
 *   '/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax.backup.0000'
 * 
 * If you experience the following error when starting a SSL secured notebook:
 *   'You must install certtool to use the secure notebook server.'
 * install net-libs/gnutls.

 * Messages for package sci-mathematics/sage-notebook-0.11.4-r2:

 * 
 * Installation of a symlink is blocked by a directory:
 *   '/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax'
 * This symlink will be merged with a different name:
 *   '/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sagenb/data/mathjax.backup.0000'
```
unmerging the notebook does not resolve things. The `mathjax` folder and symlink, `mathjax.backup.0000` remain after the unmerge and upon re-emerging the notebook yet another symlink is created
```
mathjax.backup.0001 -> /storage/strogdon/gentoo-redlizard/usr/share/mathjax
```
I'm sure this would not exist if I had removed the old notebook first, but that's not what's usually done. Joy, joy, joy!

**Comment by strogdon** [comment]
Well this may be a prefix issue. Upgraded on Gentoo from 0.11.4-r1 -> 0.11.4-r2 as part of 6.7.beta1 update and there was no problem.

**Comment by kiwifb** [comment]
No actually that's not prefix specific. I remember having it, I haven't found a way to have sagenb unmerged first. Suggestions welcome.

**Comment by strogdon** [comment]
Is it possible to have a blocker which requires manual removal first?

**Comment by kiwifb** [comment]
Yes. I think 
```
!!<=sci-mathematics/sage-notebook-0.11.4-r1
```
Will force the end user to do it themselves. Nothing can be done automatically.

**Comment by kiwifb** [comment]
Actually you only have the trouble on prefix? I remember now that I have a `pkg_preinst` section just to deal with this problem. I may have broken it when I updated the ebuild to go the ipthon's way. I replaced `ROOT` by `EROOT` in `pkg_preinst` and it is possible it broke things.

**Comment by strogdon** [comment]
Yes, change `EROOT` to `ROOT` fixed this.

**Comment by kiwifb** [comment]
So I tried to be too clever when I updated the ebuild :(

**Comment by strogdon** [comment]
I think this looks in good shape. Closing
